hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPool.java

```diff
@@ -41,7 +41,16 @@ public HiveClientPool(int poolSize, Configuration conf) {
   @Override
   protected HiveMetaStoreClient newClient()  {
     try {
-      return new HiveMetaStoreClient(hiveConf);
+      // create the metastore client based on whether we're working with Hive2 or Hive3 dependencies
+      // we need to do this because there is a breaking API change between Hive2 and Hive3
+      if (MetastoreUtil.hive3PresentOnClasspath()) {
+        return (HiveMetaStoreClient) Class
+                .forName(HiveMetaStoreClient.class.getName())
+                .getConstructor(Configuration.class)
+                .newInstance(hiveConf);
```

Comment:
Can you use `DynCtors` instead? I think this can be simpler:

```
  private static final DynConstructors.Ctor<HiveMetaStoreClient> CLIENT_CTOR = DynConstructors.builder()
      .impl(HiveMetaStoreClient.class, HiveConf.class)
      .impl(HiveMetaStoreClient.class, Configuration.class)
      .build();

  protected HiveMetaStoreClient newClient()  {
    try {
      return CLIENT_CTOR.newInstance(hiveConf);
    } catch (...) {
      ...
    }
  }
```

This also exposes a bug with the reflection path: `MetaException` is no longer thrown in the `try` block. Since that's a checked exception, it will be wrapped in a `RuntimeException`. You'll need to replace that block with `catch (RuntimeException e) { ... }` and check the cause of the exception for a `MetaException`.

Reply:
makes sense, thanks for the suggestion